### PR TITLE
Honour the `presets` and `amount` query args

### PR DESCRIPTION
**Pull request by Pomax**
closes https://github.com/mozilla/donate-wagtail/issues/844, with updated tests to ensure we don't let sneaksies through

example URL: https://donate-wagta-honor-pres-ng20bl.herokuapp.com/en-US/?presets=3.14,15.92,65.35,89.79&amount=15.92

and sorted (ascending):

https://donate-wagta-honor-pres-ng20bl.herokuapp.com/en-US/?presets=65.35,89.79,3.14,15.92&amount=15.92&sort=true

and in reverse sorting:

https://donate-wagta-honor-pres-ng20bl.herokuapp.com/en-US/?presets=3.14,15.92,65.35,89.79&amount=15.92&sort=reverse

### CODE REVIEW

File: donate/core/tests/test_models.py [hunk 1]
```diff
@@ -146,25 +146,60 @@ def test_get_initial_frequency_ignores_invalid_value(self):
             'single'
         )
 
-    def test_get_initial_currency_info_uses_arg_and_sorts(self):
-        request = RequestFactory().get('/?presets=1,9,5,3')
+    def test_get_initial_currency_info_as_specified(self):
+        request = RequestFactory().get('/?presets=2,9,5,3')
```
Comment (Pomax) [inline donate/core/tests/test_models.py]: USD minAmount is 2, so we can't use 1 in our tests anymore =)

File: donate/core/models.py [hunk 2]
```diff
@@ -72,30 +72,56 @@ def get_initial_frequency(self, request):
...
         # Check if presets have been specified in a query arg
         custom_presets = request.GET.get('presets', '').split(',')
         try:
-            custom_presets = [Decimal(p).quantize(Decimal('0.01')) for p in custom_presets]
+            min_amount = initial_currency_info.get('minAmount', 0)
+            custom_presets = [
+                amount for amount in
+                [
+                    Decimal(value).quantize(Decimal('0.01'))
+                    if float(value) >= min_amount else None
```
Comment (Pomax) [inline donate/core/models.py]: rule out things like negative amounts, zero, or anything less than the chosen currency's minimal allowed donation

File: donate/core/models.py [hunk 3]
```diff
@@ -72,30 +72,56 @@ def get_initial_frequency(self, request):
...
+                    for value in custom_presets
+                ]
+                if amount
+            ]
         except InvalidOperation:
             return initial_currency_info
+        except ValueError:
+            return initial_currency_info
```
Comment (Pomax) [inline donate/core/models.py]: short-circuit things like `1,2,foo,3`, as `foo` will make `float()` throw a ValueError

File: donate/core/models.py [hunk 4]
```diff
@@ -72,30 +72,56 @@ def get_initial_frequency(self, request):
...
 
         if not custom_presets:
             return initial_currency_info
 
-        custom_presets.sort(reverse=True)
+        if len(custom_presets) < 4:
```
Comment (Pomax) [inline donate/core/models.py]: we always want 4 options. If there's fewer presets than that, don't honour the presets.

File: donate/core/models.py [hunk 5]
```diff
@@ -72,30 +72,56 @@ def get_initial_frequency(self, request):
...
+        initial_currency_info = self.get_initial_currency_info(request, initial_currency, initial_frequency)
+        initial_amount = self.get_initial_amount(request)
         ctx.update({
             'currencies': self.currencies,
-            'initial_currency_info': self.get_initial_currency_info(
-                request, initial_currency, initial_frequency
-            ),
+            'initial_currency_info': initial_currency_info,
             'initial_frequency': initial_frequency,
+            'initial_amount': initial_amount,
```
Comment (Pomax) [inline donate/core/models.py]: used in the donation template to determine which option to preselect

File: donate/templates/fragments/donate_form.html [hunk 6]
```diff
@@ -28,7 +28,7 @@
                 <div class="donate-form__fields" id="js-donate-form-single" role="radiogroup">
                     {% for amount in initial_currency_info.presets.single %}
                     <div class="donation-amount">
-                        <input type="radio" class="donation-amount__radio" name="amount" value="{{ amount }}" id="one-time-amount-{{ forloop.counter0 }}" autocomplete="off">
+                        <input type="radio" class="donation-amount__radio" name="amount" value="{{ amount }}" id="one-time-amount-{{ forloop.counter0 }}" autocomplete="off" {% if amount == initial_amount %}checked="checked"{% endif %}>
```
Comment (cadecairos) [inline donate/templates/fragments/donate_form.html]: nice.

**Comment by WillatMozFdn** [comment]
Looks good to me @Pomax , thank you! -- let's take it live.

**Pull request merged**